src/binary/mockedRunProcess.ts

```diff
@@ -1,63 +1,79 @@
-  ReadLine
->();
-export const stdinMock: TypeMoq.IMock<stream.Writable> = TypeMoq.Mock.ofType<
-  stream.Writable
->();
-export const stdoutMock: TypeMoq.IMock<stream.Readable> = TypeMoq.Mock.ofType<
-  stream.Readable
->();
+export const requestResponseItems: Item[] = [];
+
+type CapabilitiesRequest = BinaryGenericRequest<{
+  Features: Record<string, string>;
+}>;
 
 export default function mockedRunProcess(): BinaryProcessRun {
-  spawnedProcessMock.setup((x) => x.killed).returns(() => false);
-  spawnedProcessMock.setup((x) => x.stdin).returns(() => stdinMock.object);
-  spawnedProcessMock.setup((x) => x.stdout).returns(() => stdoutMock.object);
-  setCapabilities({
-    enabled_features: [Capability.ALPHA_CAPABILITY]
-  })
+  when(spawnedProcessMock.killed).thenReturn(false);
+  when(spawnedProcessMock.stdin).thenReturn(instance(stdinMock));
+  when(spawnedProcessMock.stdout).thenReturn(instance(stdoutMock));
+  mockBinaryRequest();
+  mockCapabilitiesRequest();
+
   return {
-    proc: spawnedProcessMock.object,
-    readLine: readLineMock.object,
+    proc: instance(spawnedProcessMock),
+    readLine: instance(readLineMock),
   };
 }
-function setCapabilities(
-  capabilities: CapabilitiesResponse
-): void {
-  let requestHappened = false;
-  stdinMock.setup((x) =>
-    x.write(
-      TypeMoq.It.is<string>((request) => {
-        const capabilitiesRequest = JSON.parse(request) as { request: { Features: Record<string, unknown> } };
-        if (capabilitiesRequest?.request?.Features) {
-          requestHappened = true;
-          return true;
-        }
-
-        return false;
-      }),
-      "utf8"
-    )
+
+function mockCapabilitiesRequest() {
+  requestResponseItems.push({
+    isQualified: (request) => {
+      const completionRequest = JSON.parse(request) as CapabilitiesRequest;
+
+      return !!completionRequest?.request?.Features;
+    },
+    result: {
+      enabled_features: [Capability.ALPHA_CAPABILITY],
+    },
+  });
+}
+
+function mockBinaryRequest(): void {
+  let lineCallback: { (line: string): void } | null = null;
+
+  when(readLineMock.on("line", anyFunction())).thenCall(
+    (_: string, callback: (line: string) => void) => {
+      lineCallback = callback;
+    }
   );
-  readLineMock
-    .setup((x) => x.once("line", TypeMoq.It.isAny()))
-    .callback((x, callback: (line: string) => void) => {
-      if (!requestHappened) {
-        callback("null");
-      } else {
-        callback(JSON.stringify(capabilities));
-      }
-    });
+  when(stdinMock.write(anyString(), "utf8")).thenCall((request: string) => {
+    const matchingItem = requestResponseItems.find(({ isQualified }) =>
+      isQualified(request)
+    );
+
+    lineCallback?.(
+      matchingItem ? response(request, matchingItem.result) : "null"
+    );
+
+    return true;
+  });
 }
 
+function response(request: string, result: Result) {
+  return JSON.stringify(
+    isFunction(result) ? (result as ResultFunction)(request) : result
```

Comment:
unless i missed something, i don't see any usage of `Result` as a function, and it only seem to complicate the usage here. Just  use a simple string, and this whole function disappears.

Reply:
I used it as a function... See `setNotificationsResult` in `notification.utils.ts`

Reply:
Such wow, much useful